### ISSUE TITLE
Penny clone (Arduino-ESP32 based)

### ISSUE THREAD

**Issue by sensboston** [ssloy/penny#1]
Finally got my hands on "Penny" last weekend, here is a result: https://www.youtube.com/watch?v=iT2JZWRhM9s

ESP32 dev board a lil-bit wider for original design but it's OK. Will fork project and publish my changes soon, after "polishing" code (I hope so).

P.S. By the way, we've already met on habr. I already received PCBs but not CPUs (from ali, for original design). But I'm thinking about adopt RPi Zero W (got a couple for the sale price, $5 each).

**Comment by ssloy** [comment]
Nice job! Personally I'd like to work on the quality of the movements; right now my code produces a rough jerky pace. Ultrasonic sensors can fail to detect foam blocks, as it absorbs sound. I'd suggest to try with regular card boxes. Keep me updated please!

**Comment by sensboston** [comment]
> Ultrasonic sensors can fail to detect foam blocks

Thanks, I didn't know that (just picked some of foam specimens I brought from work, for my children), will try today different kind of blocks (children's plastic cubes).

Dmitry (btw, I'm Sergey), could you redesign frame for a little? Make a middle gap wider, legs longer and motor fixtures higher (to provide more space to fit RPi Zero and probably use 4 AA battery holder or RC battery)? Also, have you resolved issue with walking on the hardwood floor (or any other smooth surface)? My "Penny" walks nice on a carpet but fails on hardwood floor. Might be, rubber "shoes" will solve the problem but I need an idea where to get 'em.

**Comment by ssloy** [comment]
Different sensors have their own flaws, for example, infrared-based:
- cannot work outdoors (blinded by the sun)
- fail to see mirror / glass / black objects

whereas ultrasonic-based:
- work outdoors to some extent, handle mirrors, but fail to see cloth 
- must face the walls more or less orthogonally. If the angle is small enough (almost parallel to a wall), the sound bounces off the wall without returning to the receiver.

Note that I have mounted two sensors in order to detect whether the bot has to turn right or left to avoid the obstacle. The repository contains source google sketchup files, those are really easy to adjust to suit your needs.

As for the socks, for the moment I did nothing, but I am all about low budget. The possibilites are:
- bits of rubber hose ([check this photo of my quadruped](https://raw.githubusercontent.com/ssloy/quadruped/master/doc/quadruped.jpg))
- dip it into liquid rubber / silicone rubber

**Comment by sensboston** [comment]
> I have mounted two sensors in order to detect whether the bot has to turn right or left to avoid the obstacle.

Yeah, that smart. I ordered a few IR-sensors on ali, as you recommended (that was my mistake but I tried to fit in $10 budget - eBay prices from US sellers are higher 3-4 times) but they haven't arrived yet.

> The repository contains source google sketchup files, those are really easy to adjust to suit your needs.

I do have 0 experience working with google sketchup, but the main problem - I don't have 3D printer (probably, need to purchase good one, as an "entertainment" device). From my understanding, any frame changes (even a simple scaling) will require a test printing, to be sure what everything is OK. Guy from my work, Nelson, printed (for $8) my current Penny parts for me but I can't ask him (especially now, we're mostly working from home) to make a test prints.
